diff --git a/setup.py b/setup.py
index 54028463..f4e9c730 100755
--- a/setup.py
+++ b/setup.py
@@ -663,6 +663,7 @@ if __name__ == '__main__':
             c = new_compiler(verbose=1)
             distutils.sysconfig.customize_compiler(c)
             # -w suppress all warnings caused by the use of boost libraries
+            print(f'building std static library with {COMMON_MACROS=}')
             objects = c.compile(LIB_FILES,
                 include_dirs=['gsl', 'gsl/specfunc', 'build', '.', boost_include_dir] + common_extra_include_dirs,
                 output_dir='build',
@@ -671,6 +672,7 @@ if __name__ == '__main__':
             )
             c.create_static_lib(objects, "simuPOP_shstd", output_dir='build')
             #
+            print(f'building op library with {COMMON_MACROS=}')
             objects = c.compile(LIB_FILES,
                 include_dirs=['gsl', 'gsl/specfunc', 'build', '.', boost_include_dir] + common_extra_include_dirs,
                 output_dir='build',
